### ISSUE TITLE
report null coalesce or isset on undefined var array lookup

### ISSUE THREAD

**Issue by donquixote** [kalessil/phpinspectionsea#1724]
| Subject        | null coalesce on undefined var no longer reported.                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended) |
| Language level | 7.3                        |

### Code
```php
function foo() {
  if (isset($a)) {  // -> "[EA] '$a' seems to be not defined in the scope."
    return $b ?? 5;  // -> "[EA] '$a' seems to be not defined in the scope."
  }
  if (isset($c['x'])) {  // -> no report
    return $d['x'] ?? 5;  // -> no report
  }
  return NULL;
}
```

### Current behaviour
No report on undefined var array lookup.

### Expected behavior
Report undefined var array lookup. Perhaps there could be an option for this.

### Environment details
PhpStorm 2020.3.1
Build #PS-203.6682.180, built on December 30, 2020
You have a perpetual fallback license for this version.
Subscription is active until December 22, 2021.
Runtime version: 11.0.9.1+11-b1145.63 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 5.4.0-72-generic
GC: ParNew, ConcurrentMarkSweep
Memory: 1904M
Cores: 8
Registry: run.processes.with.pty=TRUE, search.everywhere.settings=true, ide.tooltip.initialDelay=952
Non-Bundled Plugins: de.thomasrosenau.diffplugin, net.seesharpsoft.intellij.plugins.csv, org.intellij.RegexpTester, org.sylfra.idea.plugins.linessorter, R4Intellij, Dart, lv.midiana.misc.idea-plugins.deep-js-completion, com.dmarcotte.handlebars, com.kalessil.phpStorm.phpInspectionsEA, de.espend.idea.php.annotation, fr.adrienbrault.idea.symfony2plugin, de.espend.idea.php.drupal, ru.adelf.idea.dotenv
Current Desktop: MATE

### Side note
I had a look into the fixtures in `isset-operations-variable-existence.php`
```php
            return [ $x ?? $y, isset($x), empty($x) ];
```
and 
```php
        $x = <error descr="[EA] '$x' seems to be not defined in the scope.">$x</error> ?? '';
```

Having those 3 statements in one line, would this not obfuscate a bug where only one of the three is reported?
Perhaps better like this?
```php
            return [ $x0 ?? $y, isset($x1), empty($x2) ];
```
and 
```php
        $x = <error descr="[EA] '$x0' seems to be not defined in the scope.">$x0</error> ?? '';
        $x = <error descr="[EA] '$x1' seems to be not defined in the scope.">$x1</error> ?? '';
        $x = <error descr="[EA] '$x2' seems to be not defined in the scope.">$x2</error> ?? '';
```

**Comment by ktomk** [comment]
I get all variables from your example marked by EA, all with the correct variable names, e.g.

> [EA] '$c' seems to be not defined in the scope.

So this is perhaps already fixed, try the updated version of the plugin to verify.

**Comment by donquixote** [comment]
> So this is perhaps already fixed, try the updated version of the plugin (and maybe latest Phpstorm EAP) to verify.
I just updated:

> PhpStorm 2021.1.2
> Build #PS-211.7142.44, built on April 30, 2021
> Runtime version: 11.0.10+9-b1341.41 amd64
> VM: Dynamic Code Evolution 64-Bit Server VM by JetBrains s.r.o.
> Linux 5.4.0-73-generic
> GC: ParNew, ConcurrentMarkSweep
> Memory: 1904M
> Cores: 8
> Registry: run.processes.with.pty=TRUE, search.everywhere.settings=true, ide.tooltip.initialDelay=952
> Non-Bundled Plugins: net.seesharpsoft.intellij.plugins.csv (2.16.3), org.intellij.RegexpTester (1.0.8), org.sylfra.idea.plugins.linessorter (1.0.1), R4Intellij (211.6693.2), Dart (211.7233), lv.midiana.misc.idea-plugins.deep-js-completion (2021.04.21.002), com.dmarcotte.handlebars (211.6693.44), com.kalessil.phpStorm.phpInspectionsEA (4.0.6.3), de.espend.idea.php.annotation (8.0.0), fr.adrienbrault.idea.symfony2plugin (0.23.209), de.espend.idea.php.drupal (0.5.1), ru.adelf.idea.dotenv (2021.1)
> Current Desktop: MATE

Also created a new project which only contains this one file.
Set PHP version to 7.3.
Still the problem is not being reported.

> and maybe latest Phpstorm EAP


> why not consider to upgrade?
that's a noble idea.. but does this explain the difference?

**Comment by donquixote** [comment]
> why not consider to upgrade?

This is quite interesting..
I found https://kalessil.github.io/php-inspections-ultimate.html#pricing-mark, but nothing is clickable here.
I did find EA Ultimate through PhpStorm itself, and installed it, disabled EA Extended. But I don't see any place where I am asked to pay something.
It seems I now have a limited subset of features, and it seems to be less than what I get with EA Extended.

It seems Vladimir does not want my money :)

**Comment by ktomk** [comment]
> that's a noble idea.. but does this explain the difference?

well I would just support a project I use every day, but that is entirely your decision. there was nothing noble on my end actually, but it's nice if you make me feel so.

however: I've disabled ultimate and re-installed extended (which I've been using until a couple of months ago for too long) and updated phpstorm  (from EAP to the newer preview edition which just came out):

the reporting indeed _is_ different.

For example in Phpstorm on a PHP 5.3 language level and this code:

~~~
<?php

function foo() {
    if (isset($a)) {  // [EA] '$a' seems to be not defined in the scope. 
        return $b ?: 5;  // Undefined variable '$b' 
    }
    if (isset($c['x'])) {  // <no report>
        return $d['x'] ?: 5;  // Undefined variable '$d' 
    }
    return NULL;
}
~~~

which show that the ultimate edition detects more of the undefined variables. in extended it is only $a, the other two are from phpstorm. and ultimate did detect $c as well, which gets undetected for phpstorm.

double checked after switching back to ultimate, after this little comparison its likely clear why, right?

**Comment by ktomk** [comment]
> It seems Vladimir does not want my money :)

Please see: https://www.patreon.com/posts/how-to-obtain-40038459 and consider some time, this is not automated purchasing.

**Comment by donquixote** [comment]
> well I would just support a project I use every day, but that is entirely your decision..

I did it! I chose the cheapo option for now, but perhaps will change it.
In my interactions in this issue queue I always had the impression that the maintainer does a really good job, so he deserves my money. But probably would be a good idea to mention the Patreon on https://kalessil.github.io/php-inspections-ultimate.html#pricing-mark, which is the first page people will find when looking to get a license.

**Comment by donquixote** [comment]
I finally have the "Ultimate" version of the plugin.
Now all of the problems are reported correctly.

Some feedback:
- When using the free version (EA Extended), it would be useful to hint at features you get with the full version.
- In the marketplace we now find two versions of EA Ultimate. One of them is "maintenance only", but it is hard to distinguish them in a list. Perhaps the old one should be renamed to something like "** (deprecated)" or "** (old version)".

**Comment by ktomk** [comment]
@donquixote the last point might be due to recent changes as I have not seen it before. good point thought. I've also seen there was very recent update. but I'm also switching on EAP versions of Phpstorm itself so there could be some room required to have multiple version. As I'm often using EAP I won't complain as so far its well supported by EA inspections ;)

**Comment by ea-inspections-team** [comment]
When I remember correctly, the inspection is on of enhanced from Ultimate.
@donquixote : congrats and thanks for supporting the project.

Note for future us: revisit implementations to clarify which patterns supported in both versions.

**Comment by donquixote** [comment]
> When I remember correctly, the inspection is on of enhanced from Ultimate.

That's a confusing sentence :)
I found 3 versions of the inspection discussed here:
- The old version (*) of EA Ultimate has the worst version of this inspection, where none of the problems in the code snippet are found. Or perhaps this is because I did not add a license?
- The regular EA Extended has a medium version where half of the problems are found.
- The new EA Ultimate has the best version of this inspection, where all problems are found.

(*) By "old" I mean the one available on https://plugins.jetbrains.com/plugin/10215-php-inspections-ea-ultimate- which is marked as deprecated. By "new" I mean the one available on https://plugins.jetbrains.com/plugin/16935-php-inspections-ea-ultimate-. Interestingly, the "old" version says it is 4.91, whereas the "new" version says it is 4.59. Super confusing :)

**Comment by ea-inspections-team** [comment]
> The regular EA Extended has a medium version where half of the problems are found.
That is correct.

> The old version (*) of EA Ultimate has the worst version of this inspection, where none of the problems in the code snippet are found. Or perhaps this is because I did not add a license?

Likely it was the case. Functionality-wise the both new and old EA Ultimate are identical.

**Comment by ea-inspections-team** [comment]
Closing: we have re-checked the original code fragment - IDE + EA Ultimate are covering all four cases.
